### ISSUE TITLE
[Feature Request] Improving examples scripts

### ISSUE THREAD

**Issue by BHZ-BER** [radixark/miles#315]
Right now, a lot of training examples are based on thinking models with default `rollout-max-response-len` and `eval-max-response-len`. This leads to a lot of truncations and a lot of examples ended up only learning to shorten the responses without really improving the performance.

We should consider switching to a better, more realistic examples in [quick_start](https://github.com/radixark/miles/blob/main/docs/en/get_started/quick_start.md). We can start from either Qwen 2.5 Instruct, or Qwen 3 4B base. And do RL with DAPO 17k. 

Similarly, we need the same changes for `scripts/run_qwen3_30b_a3b.py`, `scripts/run_glm45_355b_a32b.py` etc., and also provide better documentations there.

**Comment by KennyYao2001** [comment]
would love to contribute!

**Comment by gongyisheng** [comment]
I want to contribute on this as well

**Comment by gongyisheng** [comment]
@BHZ-BER I notice that miles has two folders with run example scripts, `examples/` and `scripts/`, based on my understanding we need to update `scripts/` in this issue, is it correct?

**Comment by fzyzcjy** [comment]
> based on my understanding we need to update scripts/ in this issue

Yes it is in scripts folder. I think the first step may be find such an example w/ (significantly) increasing accuracy.

**Comment by BHZ-BER** [comment]
@gongyisheng @fzyzcjy Yes, and also it's best to include the best example in the [quick start doc here](https://github.com/radixark/miles/blob/main/docs/en/get_started/quick_start.md).

**Comment by zhaochenyang20** [comment]
Hey Team, I already called out the community possibly on this. At the same time, I can be the example reviewer.

**Comment by KennyYao2001** [comment]
> Hey Team, I already called out the community possibly on this. At the same time, I can be the example reviewer.

Thanks! Is there an existing public dashboard or doc that tracks the key metrics for these runs (e.g., AIME eval results, truncation rate, rollout/eval token stats)? If not, what metrics and reporting format would be most valuable?

**Comment by zhaochenyang20** [comment]
Thanks, it really makes sense to me. @KennyYao2001 To be honest, I also do not know the answer. Could you possibly share your wandb link, and let's pick out of it.

**Comment by KennyYao2001** [comment]
Sure! The wandb link is [here](https://wandb.ai/kenny_yao-carnegie-mellon-university/miles-dev/runs/b429qvir?nw=nwuserkenny_yao).

The args I was using for training qwen3-4B is the following：

```#!/bin/bash

# for rerun the task
pkill -9 sglang
sleep 3
ray stop --force
pkill -9 ray
pkill -9 python
sleep 3
pkill -9 ray
pkill -9 python

set -ex

# will prevent ray from buffering stdout/stderr
export PYTHONBUFFERED=16

NVLINK_COUNT=$(nvidia-smi topo -m 2>/dev/null | grep -o 'NV[0-9][0-9]*' | wc -l)
if [ "$NVLINK_COUNT" -gt 0 ]; then
    HAS_NVLINK=1
else
    HAS_NVLINK=0
fi
echo "HAS_NVLINK: $HAS_NVLINK (detected $NVLINK_COUNT NVLink references)"

SCRIPT_DIR="$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)"
source "${SCRIPT_DIR}/models/qwen3-4B.sh"

CKPT_ARGS=(
   --hf-checkpoint /root/Qwen3-4B
   #--hf-checkpoint /root/Qwen3-4B-FP8
   --ref-load /root/Qwen3-4B_torch_dist
   --load /root/Qwen3-4B_miles/
   --save /root/Qwen3-4B_miles/
   --save-interval 20
)

ROLLOUT_ARGS=(
   --prompt-data /root/dapo-math-17k/dapo-math-17k.jsonl
   --input-key prompt
   --label-key label
   --apply-chat-template
   --rollout-shuffle
   --rm-type deepscaler
   --num-rollout 3000
   --rollout-batch-size 32
   --n-samples-per-prompt 8
   --rollout-max-response-len 8192
   --rollout-temperature 0.8

   --global-batch-size 256
   --balance-data
)

EVAL_ARGS=(
   --eval-interval 20
   --eval-prompt-data aime /root/aime-2024/aime-2024.jsonl
   --n-samples-per-eval-prompt 16
   --eval-max-response-len 16384
   --eval-top-p 0.7
)

PERF_ARGS=(
   --tensor-model-parallel-size 2
   --sequence-parallel
   --pipeline-model-parallel-size 1
   --context-parallel-size 1
   --expert-model-parallel-size 1
   --expert-tensor-parallel-size 1

   --recompute-granularity full
   --recompute-method uniform
   --recompute-num-layers 1

   # --micro-batch-size 1
   --use-dynamic-batch-size
   --max-tokens-per-gpu 9216
)

GRPO_ARGS=(
   --advantage-estimator grpo
   --use-kl-loss
   --kl-loss-coef 0.00
   --kl-loss-type low_var_kl
   --entropy-coef 0.00
   --eps-clip 0.2
   --eps-clip-high 0.28
)

OPTIMIZER_ARGS=(
   --optimizer adam
   --lr 1e-6
   --lr-decay-style constant
   --weight-decay 0.1
   --adam-beta1 0.9
   --adam-beta2 0.98
)

WANDB_ARGS=(
   --use-wandb
   --wandb-project miles-dev
   --wandb-group qwen3-4B-test
   --wandb-key ${WANDB_KEY}
)

SGLANG_ARGS=(
   --rollout-num-gpus-per-engine 1
   --sglang-mem-fraction-static 0.7
)

MISC_ARGS=(
   # default dropout in megatron is 0.1
   --attention-dropout 0.0
   --hidden-dropout 0.0
   # should be good for model performance
   --accumulate-allreduce-grads-in-fp32
   --attention-softmax-in-fp32
   # need to comment this when using model with MLA
   --attention-backend flash
)

# launch the master node of ray in container
export MASTER_ADDR=${MASTER_ADDR:-"127.0.0.1"}
ray start --head --node-ip-address ${MASTER_ADDR} --num-gpus 2 --disable-usage-stats --dashboard-host=0.0.0.0 --dashboard-port=8265

# Build the runtime environment JSON with proper variable substitution
RUNTIME_ENV_JSON="{
  \"env_vars\": {
    \"PYTHONPATH\": \"/root/Megatron-LM/\",
    \"CUDA_DEVICE_MAX_CONNECTIONS\": \"1\",
    \"NCCL_NVLS_ENABLE\": \"${HAS_NVLINK}\"
  }
}"

ray job submit --address="http://127.0.0.1:8265" \
   --runtime-env-json="${RUNTIME_ENV_JSON}" \
   -- python3 train.py \
   --actor-num-nodes 1 \
   --actor-num-gpus-per-node 2 \
   --colocate \
   ${MODEL_ARGS[@]} \
   ${CKPT_ARGS[@]} \
   ${ROLLOUT_ARGS[@]} \
   ${OPTIMIZER_ARGS[@]} \
   ${GRPO_ARGS[@]} \
   ${WANDB_ARGS[@]} \
   ${PERF_ARGS[@]} \
   ${EVAL_ARGS[@]} \
   ${SGLANG_ARGS[@]} \
   ${MISC_ARGS[@]}```
```
Particularly highlighting the 
```
--rollout-max-response-len 8192
--eval-max-response-len 16384
```

In this settings, the truncation radio in rollout is the around 0.5 at the beginning.

**Comment by zhaochenyang20** [comment]
> Sure! The wandb link is [here](https://wandb.ai/kenny_yao-carnegie-mellon-university/miles-dev/runs/b429qvir?nw=nwuserkenny_yao).
> 
> The args I was using for training qwen3-4B is the following：
> 
> ```
> 
> # for rerun the task
> pkill -9 sglang
> sleep 3
> ray stop --force
> pkill -9 ray
> pkill -9 python
> sleep 3
> pkill -9 ray
> pkill -9 python
> 
> set -ex
> 
> # will prevent ray from buffering stdout/stderr
> export PYTHONBUFFERED=16
> 
> NVLINK_COUNT=$(nvidia-smi topo -m 2>/dev/null | grep -o 'NV[0-9][0-9]*' | wc -l)
> if [ "$NVLINK_COUNT" -gt 0 ]; then
>     HAS_NVLINK=1
> else
>     HAS_NVLINK=0
> fi
> echo "HAS_NVLINK: $HAS_NVLINK (detected $NVLINK_COUNT NVLink references)"
> 
> SCRIPT_DIR="$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)"
> source "${SCRIPT_DIR}/models/qwen3-4B.sh"
> 
> CKPT_ARGS=(
>    --hf-checkpoint /root/Qwen3-4B
>    #--hf-checkpoint /root/Qwen3-4B-FP8
>    --ref-load /root/Qwen3-4B_torch_dist
>    --load /root/Qwen3-4B_miles/
>    --save /root/Qwen3-4B_miles/
>    --save-interval 20
> )
> 
> ROLLOUT_ARGS=(
>    --prompt-data /root/dapo-math-17k/dapo-math-17k.jsonl
>    --input-key prompt
>    --label-key label
>    --apply-chat-template
>    --rollout-shuffle
>    --rm-type deepscaler
>    --num-rollout 3000
>    --rollout-batch-size 32
>    --n-samples-per-prompt 8
>    --rollout-max-response-len 8192
>    --rollout-temperature 0.8
> 
>    --global-batch-size 256
>    --balance-data
> )
> 
> EVAL_ARGS=(
>    --eval-interval 20
>    --eval-prompt-data aime /root/aime-2024/aime-2024.jsonl
>    --n-samples-per-eval-prompt 16
>    --eval-max-response-len 16384
>    --eval-top-p 0.7
> )
> 
> PERF_ARGS=(
>    --tensor-model-parallel-size 2
>    --sequence-parallel
>    --pipeline-model-parallel-size 1
>    --context-parallel-size 1
>    --expert-model-parallel-size 1
>    --expert-tensor-parallel-size 1
> 
>    --recompute-granularity full
>    --recompute-method uniform
>    --recompute-num-layers 1
> 
>    # --micro-batch-size 1
>    --use-dynamic-batch-size
>    --max-tokens-per-gpu 9216
> )
> 
> GRPO_ARGS=(
>    --advantage-estimator grpo
>    --use-kl-loss
>    --kl-loss-coef 0.00
>    --kl-loss-type low_var_kl
>    --entropy-coef 0.00
>    --eps-clip 0.2
>    --eps-clip-high 0.28
> )
> 
> OPTIMIZER_ARGS=(
>    --optimizer adam
>    --lr 1e-6
>    --lr-decay-style constant
>    --weight-decay 0.1
>    --adam-beta1 0.9
>    --adam-beta2 0.98
> )
> 
> WANDB_ARGS=(
>    --use-wandb
>    --wandb-project miles-dev
>    --wandb-group qwen3-4B-test
>    --wandb-key ${WANDB_KEY}
> )
> 
> SGLANG_ARGS=(
>    --rollout-num-gpus-per-engine 1
>    --sglang-mem-fraction-static 0.7
> )
> 
> MISC_ARGS=(
>    # default dropout in megatron is 0.1
>    --attention-dropout 0.0
>    --hidden-dropout 0.0
>    # should be good for model performance
>    --accumulate-allreduce-grads-in-fp32
>    --attention-softmax-in-fp32
>    # need to comment this when using model with MLA
>    --attention-backend flash
> )
> 
> # launch the master node of ray in container
> export MASTER_ADDR=${MASTER_ADDR:-"127.0.0.1"}
> ray start --head --node-ip-address ${MASTER_ADDR} --num-gpus 2 --disable-usage-stats --dashboard-host=0.0.0.0 --dashboard-port=8265
> 
> # Build the runtime environment JSON with proper variable substitution
> RUNTIME_ENV_JSON="{
>   \"env_vars\": {
>     \"PYTHONPATH\": \"/root/Megatron-LM/\",
>     \"CUDA_DEVICE_MAX_CONNECTIONS\": \"1\",
>     \"NCCL_NVLS_ENABLE\": \"${HAS_NVLINK}\"
>   }
> }"
> 
> ray job submit --address="http://127.0.0.1:8265" \
>    --runtime-env-json="${RUNTIME_ENV_JSON}" \
>    -- python3 train.py \
>    --actor-num-nodes 1 \
>    --actor-num-gpus-per-node 2 \
>    --colocate \
>    ${MODEL_ARGS[@]} \
>    ${CKPT_ARGS[@]} \
>    ${ROLLOUT_ARGS[@]} \
>    ${OPTIMIZER_ARGS[@]} \
>    ${GRPO_ARGS[@]} \
>    ${WANDB_ARGS[@]} \
>    ${PERF_ARGS[@]} \
>    ${EVAL_ARGS[@]} \
>    ${SGLANG_ARGS[@]} \
>    ${MISC_ARGS[@]}```
> ```
> 
> Particularly highlighting the
> 
> ```
> --rollout-max-response-len 8192
> --eval-max-response-len 16384
> ```
> 
> In this settings, the truncation radio in rollout is the around 0.5 at the beginning.

Thanks. Please add zhaochenyang20@gmail.com to your wandb project. Then I can see. 😂

**Comment by zhaochenyang20** [comment]
@KennyYao2001 Also, please see the description of this issue provided by @BHZ-BER What do you think of it?

**Comment by zhaochenyang20** [comment]
@KennyYao2001 could you please add my wechat 18015766633

**Comment by KennyYao2001** [comment]
> Also, please see the description of this issue provided by [@BHZ-BER](https://github.com/BHZ-BER) What do you think of it?

I think the main goal is to bring the truncation rate at the very beginning of training below 10%, to prevent the model from learning a length-avoidance shortcut.

We currently see two directions:
	Option A.	Parameter-side approach:
Increase rollout-max-response-len, while slightly reducing throughput (e.g., smaller batch size or fewer samples per prompt) to stay within memory limits. This allows early rollouts to finish properly and receive meaningful rewards instead of being truncated.
	Option B.	Data-side approach:
Start rollouts with simpler, shorter questions that naturally require fewer tokens, and then gradually introduce longer problems while relaxing rollout-max-response-len over time.

**Comment by gongyisheng** [comment]
@BHZ-BER @zhaochenyang20 Discussed with @KennyYao2001 , based on my past experience and recent [research paper](https://arxiv.org/abs/2512.12895), the example settings (train rollout temperature=0.8, eval rollout top_p=0.7) can cause degeneration problems, which result in high truncation ratio. I will test the case that setting train/eval temperature to 1, and top_p=1 to see if it helps.

**Comment by gongyisheng** [comment]
Besides, I'm running qwen3-30b-a3b example on 8XH100, but it raises CudaOOM error during training (crash in the middle, several minutes after start). Have anyone tried this example script yet and know the best way to fix?

**Comment by gongyisheng** [comment]
Regarding Qwen3-30B-A3B case, I did controlled experiments on potential methods to resolve the issue
- baseline: settings of current example script: Qwen3-30B-A3B model,  rollout-max-resp-len=8192, temperature=0.8, enable think
- temperature_1: set `--rollout-temperature 1`
- max_len_16384: set `--rollout-max-response-len 16384`
- disable_think: add `--apply-chat-template-kwargs '{"enable_thinking":false}'`
- instruct_model: use instruct model `Qwen3-30B-A3B-Instruct-2507`

I run experiments on 8x6000 pro instance, 8xH100 encounters OOM error, I will investigate it later. 
I stop experiment after running for 10 steps, focus on truncation ratio at start. 

results:
| setting            | avg truncation ratio | avg response length | avg raw reward |
|--------------------|----------------------|--------------------|----------------|
| baseline           | 0.4980               | 6798.03            | 0.4918         |
| temperature_1    | 0.4930               | 6793.87            | 0.4969         |
| max_len_16384      | 0.1480               | 9151.26            | 0.8012         |
| disable_think      | 0.0027               | 2041.95            | 0.5561         |
| instruct model     | 0.1484               | 4018.35            | 0.7944         |

screenshots:

<img width="402" height="310" alt="Image" src="https://github.com/user-attachments/assets/d2def232-a176-4a28-b39f-530dd4b2004a" />

conclusion:
- choose disable think or instruct model, we can document it in quick start guide. 
- set temperature = 1 during train & eval can be a improvement suggested by paper.

let me know your idea

**Comment by zhaochenyang20** [comment]
Wow. Nice done! @gongyisheng

**Comment by zhaochenyang20** [comment]
Let me bring to Banghua to see his feedbacks.

**Comment by BHZ-BER** [comment]
Disable think is good idea. I think Qwen 3 30BA3B might still be a bit too large for people to start with. Maybe we can try Qwen2.5-3B-Instruct or Qwen2.5-7B-Instruct? (And we can also include the Qwen3 30BA3B example after that if we want to demonstrate a small MoE training example

**Comment by KennyYao2001** [comment]
> Disable think is good idea. I think Qwen 3 30BA3B might still be a bit too large for people to start with. Maybe we can try Qwen2.5-3B-Instruct or Qwen2.5-7B-Instruct? (And we can also include the Qwen3 30BA3B example after that if we want to demonstrate a small MoE training example

For the Qwen3-4B example case, I did some experiment sync with @gongyisheng  as well.
* baseline: settings of current example script, rollout-max-resp-len=8192, temperature=0.8, enable thinking.
* temperature_1: set --rollout-temperature 1
* max_len_16384: set --rollout-max-response-len 16384
* temperature_1 with max_len_16384: set --rollout-temperature 1 and set --rollout-max-response-len 16384
* disable_think: add --apply-chat-template-kwargs '{"enable_thinking":false}'
* base_model: use base model Qwen3-30B-A3B-Instruct-2507

<meta charset="utf-8"><div data-page-id="CqShdUF4BotwEPx9MDMlQ8p3glf" data-lark-html-role="root" data-docx-has-block-data="true"><div>
Setting | avg rollout truncation ratio | avg rollout raw reward | eval / truncation_ratio | eval / AIME
-- | -- | -- | -- | --
Default Script | 0.42482 | 0.54492 | 0.32292 | 0.56911
temperature_1 | 0.43491 | 0.57578 | 0.30277 | 0.62008
max_len_16384 | 0.15378 | 0.77891 | 0.34792 | 0.60072
max_len_16384 + temperature_1 | 0.15204 | 0.77656 | 0.34375 | 0.56667
Base_model | 0.03516 | 0.00000 | 0.18333 | 0.00000
disable_think | 0.00314 | 0.00000 | 0.21850 | 0.00000

<img width="1968" height="1146" alt="Image" src="https://github.com/user-attachments/assets/d75f752d-ccb9-4f0c-aaf9-eeb79e1748af" />

<img width="1972" height="1138" alt="Image" src="https://github.com/user-attachments/assets/7add1ca2-bb5d-4fed-b425-2b54fc583a98" />

Conclusion: 
- Sample a few responses from default script (https://chatgpt.com/share/69443e13-643c-800c-82ed-2d77789eae2f) from rollout and response contain serve degeneration problems. This is one of the important reasons for truncation. Current  metric rollout/repetition_frac is not good at detecting degeneration (see below image). I can investigate to better measure degeneration.
  - rollout/repetition_frac is calculated based on 1) length > 10,000 2) s[:-10000] compression ratio > 10.
- disable_think or use base/instruction model will lead to eval/AIME to 0, might be a potential bug
